synapse/storage/databases/main/events.py

```diff
@@ -1795,6 +1795,10 @@ def _handle_event_relations(
             txn.call_after(
                 self.store.get_thread_summary.invalidate, (parent_id, event.room_id)
             )
+            txn.call_after(
+                self.store.get_thread_participated.invalidate,
+                (parent_id, event.room_id),
+            )
```

Comment:
Is this invalidation always necessary?

If we're responding to a thread for the first time, I can see that we need an invalidation here. (participated: false -> participated: true).

Is it ever possible to transition from participated: true to participated: false after your first contribution to a thread?

My first thought was "surely not", but then I thought about
- redacting/deleting events
- being banned before you wrote in the thread, but only discovering that later after a netsplit + state res

and I wasn't so sure.

Reply:
I had the same thought process as you, but I don't think we have an easy way to invalidate-on-condition. I guess this does just call a method though so I could probably do something like:

```suggestion
            def maybe_invalidate(parent_id, room_id):
                if self.store.get_thread_participated.cache.get((parent_id, room_id)) is False:
                	self.store.get_thread_participated.invalidate(parent_id, room_id)
            txn.call_after(
                maybe_invalidate,
                (parent_id, event.room_id),
            )
```

Although this is making realize that this cache needs to be `tree=True`, I think. Since we invalidate it with a partial key.

Reply:
I'd be happy with just invalidating it anyway TBH (and it means we wouldn't have to worry about the other transitions above).

> Although this is making realize that this cache needs to be tree=True, I think. Since we invalidate it with a partial key.

I agree with this reasoning; but I'm slightly surprised why we can't invalidate just a single entry. Do we have access to the sender of the event being processed here?

Reply:
We have the entire event, so yeah you're right! 👍  Thanks!